lib/search/fragmentation_scheme.ex

```diff
@@ -0,0 +1,73 @@
+defmodule Search.FragmentationScheme do
+  @doc """
+  Splits a binary into multiple binaries that satisfy limitations specified by opts.
+
+  If possible, splits the text on whitespace to preserve words. If that is impossible, splits text in between graphemes.
+
+  Supported options:
+
+  * `:max_size` - maximum byte_size of the output binaries. The output binaries may have size less or equal to that
+    value, which also should guarantee the sequence length after tokenization will be bounded by this value.
+  """
+  def split(text, opts \\ [])
+  def split("", _opts), do: []
+
+  def split(text, opts) when is_binary(text) do
+    case Keyword.get(opts, :max_size) do
+      nil -> [text]
+      max_size -> do_split(text, [], max_size)
+    end
+  end
+
+  def recombine(chunks), do: Enum.join(chunks)
+
+  defp do_split("", acc, _max_size), do: Enum.reverse(acc)
+
+  defp do_split(text, acc, max_size) do
+    # capture the next word along with trailing whitespace and leading whitespace, if any
+    [next_word] =
+      Regex.run(~r/^([\s]*[^\s]+\s+)[^\s]*/, text, capture: :all_but_first) ||
+        [text]
+
+    word_chunks =
+      if byte_size(next_word) > max_size do
+        split_word("", next_word, [], max_size)
+      else
+        [next_word]
+      end
+
+    next_text = binary_slice(text, byte_size(next_word)..-1//1)
+
+    case {word_chunks, acc} do
+      {_, []} ->
+        do_split(next_text, word_chunks, max_size)
+
+      {[single_word], [acc_head | acc_tail]} ->
+        # we can try extending the last word in accumulator
+        if byte_size(acc_head) + byte_size(single_word) <= max_size do
+          do_split(next_text, [acc_head <> single_word | acc_tail], max_size)
```

Comment:
Here we are incrementally building the binaries, but ultimately what we want to do is *split* a binary into adjacent chunks, so it would be more memory-efficient to accumulate chunks as `{offset, size}` and then produce the corresponding slices at the end.

Also, this could be a recursion using `String.next_grapheme/2` to iterate over all graphemes (avoiding `Regex.run`). One way to do it would be to track the current string, offset, chunk start position and previous word end position. This is a less important optimisation, but it may actually end up simpler, so I am just dropping the idea :)

Reply:
Did as advised, and the function is indeed simpler, and I even discovered another edge case to be covered while redoing the it. The function seems rather ugly still though to me, so if you have any further suggestions about it I'd be more than happy to hear them